quickwit/quickwit-control-plane/src/ingest/ingest_controller.rs

```diff
@@ -254,132 +319,96 @@ impl IngestController {
+        // so the control plane is not necessarily aware that they are closed.
+        //
+        // Router can tip the cp about which shard is closed, so that we can update our
+        // internal state.
         self.handle_closed_shards(get_open_shards_request.closed_shards, model);
 
+        let num_subrequests = get_open_shards_request.subrequests.len();
+        let mut get_or_create_open_shards_successes = Vec::with_capacity(num_subrequests);
+        let mut get_or_create_open_shards_failures = Vec::new();
+
+        let mut num_missing_shards_per_source_uids = HashMap::new();
+
         let unavailable_leaders: FnvHashSet<NodeId> = get_open_shards_request
             .unavailable_leaders
             .into_iter()
-            .map(|ingester_id| ingester_id.into())
+            .map(Into::into)
             .collect();
 
-        let num_subrequests = get_open_shards_request.subrequests.len();
-        let mut get_or_create_open_shards_successes = Vec::with_capacity(num_subrequests);
-        let mut get_or_create_open_shards_failures = Vec::new();
-        let mut open_shards_subrequests = Vec::new();
-
-        for get_open_shards_subrequest in get_open_shards_request.subrequests {
-            let Some(index_uid) = model.index_uid(&get_open_shards_subrequest.index_id) else {
-                let get_or_create_open_shards_failure = GetOrCreateOpenShardsFailure {
-                    subrequest_id: get_open_shards_subrequest.subrequest_id,
-                    index_id: get_open_shards_subrequest.index_id,
-                    source_id: get_open_shards_subrequest.source_id,
-                    reason: GetOrCreateOpenShardsFailureReason::IndexNotFound as i32,
+        // We do a first pass to identify the shards that are missing from the model and need to be
+        // created.
+        for get_open_shards_subrequest in &get_open_shards_request.subrequests {
+            if let Ok(None) =
+                get_open_shard_from_model(get_open_shards_subrequest, model, &unavailable_leaders)
+            {
+                // We did not find any open shard in the model, we will have to create one.
+                // Let's keep track of all of the source that require new shards, so we can batch
+                // create them after this loop.
+                let index_uid = model
+                    .index_uid(&get_open_shards_subrequest.index_id)
+                    .unwrap();
+                let source_uid = SourceUid {
+                    index_uid,
+                    source_id: get_open_shards_subrequest.source_id.clone(),
                 };
-                get_or_create_open_shards_failures.push(get_or_create_open_shards_failure);
-                continue;
-            };
-            let Some(open_shard_entries) = model.find_open_shards(
-                &index_uid,
-                &get_open_shards_subrequest.source_id,
+                num_missing_shards_per_source_uids
+                    .entry(source_uid)
```

Comment:
nit: `*foo.entry().or_default += 1` seems more natural.

Reply:
the pattern was offered to you by chatgpt :)